src/main.rs

```diff
@@ -3,23 +3,112 @@
+    .global init_rcp
+    init_rcp:
+        // Enable the RCP.  To save space, it is assumed that no other
+        // coprocessors are enabled.
+        ldr r0, =0xe0000000 + 0x0000ed88  // PPB_BASE + M33_CPACR_OFFSET
+        movs r1, 0x0000c000 // CPACR_CP7_BITS
+        str r1, [r0]
+
+        // Check that to see if the RCP is already initialized.
+        //
+        // Since this check requires passing `r15` to `mrc` and the inline
+        // assembler will not allow this, we hard code the instruction here:
+        //     `mrc p7, #1, r15, c0, c0, #0`
+        .byte 0x30
+        .byte 0xee
+        .byte 0x10
+        .byte 0xf7
+
+        // Skip initialization if already initialized.
+        bmi 2f
+
+        // Initialize the RCP.
+        mcrr p7, #8, r0, r0, c0
+        mcrr p7, #8, r0, r0, c1
+
+        // Signal other core.
+        sev
+
+        2:
+        bx lr
+    "#,
+);
+
+unsafe fn lookup_func_v1(tag: u32) -> usize {
     type RomTableLookupFn = unsafe extern "C" fn(table: *const u16, code: u32) -> usize;
-    /// This location in flash holds a 16-bit truncated pointer for the ROM lookup function
+
+    /// This location in ROM holds a 16-bit truncated pointer for the ROM lookup function for
+    /// v2 ROMs.
     const ROM_TABLE_LOOKUP_PTR: *const u16 = 0x0000_0018 as _;
-    /// This location in flash holds a 16-bit truncated pointer for the ROM function table
+
+    /// This location in ROM holds a 16-bit truncated pointer for the ROM function table
     /// (there's also a ROM data table which we don't need)
     const FUNC_TABLE: *const u16 = 0x0000_0014 as _;
+
+    let lookup_func = ROM_TABLE_LOOKUP_PTR.read() as usize;
+    let lookup_func: RomTableLookupFn = core::mem::transmute(lookup_func);
+    let table = FUNC_TABLE.read() as usize;
+    lookup_func(table as *const u16, tag)
+}
+
+unsafe fn lookup_func_v2(tag: u32) -> usize {
+    type RomTableLookupFn = unsafe extern "C" fn(code: u32, mask: u32) -> usize;
+
+    /// This location in ROM holds a 16-bit truncated pointer for the ROM lookup function for
+    /// V2 ROMs.
+    const ROM_TABLE_LOOKUP_PTR: *const u16 = 0x0000_0016 as _;
+
+    /// The flash-algo needs to run in secure mode so we need too look up
+    /// functions in that context
+    const RT_FLAG_FUNC_ARM_SEC: u32 = 0x0004;
+
+    // The RCP (redundancy coprocessor) must be enabled in order to call ROM functions.
+    init_rcp();
+
+    let lookup_func = ROM_TABLE_LOOKUP_PTR.read() as usize;
+    let lookup_func: RomTableLookupFn = core::mem::transmute(lookup_func);
+    lookup_func(tag, RT_FLAG_FUNC_ARM_SEC)
+}
+
+fn find_func<T>(tag: [u8; 2]) -> Result<T, ErrorCode> {
+    let tag = u16::from_le_bytes(tag) as u32;
+
+    /// This location in ROM holds a 3 byte magic value that confirms the validity of the
+    /// ROM as well as identifying different interfaces (referred here in as V1 and V2).
+    const BOOTROM_MAGIC: *const [u8; 3] = 0x0000_0010 as _;
+
+    /// Magic value for V1 ROMs.
```

Comment:
Do you mean RP2040, or RP2350 A1?

Reply:
I mean RP2040 here but struggled with the right nomenclature.  The RP2040 datasheet defines the magic as `'M', 'u', 0x01` and  2350 datasheet defines the magic as `'M', 'u', 0x02`.  This is different from the byte at `0x13` which is the "bootrom" version.   I'm happy to change these to RP2040 and RP235x if that's clearer.

Reply:
I'd call them RP2040 and RP2350 I think. Especially as each ROM has multiple revisions.

Reply:
renamed